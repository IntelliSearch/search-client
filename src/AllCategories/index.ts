export * from './AllCategoriesSettings';
export * from './AllCategories';
