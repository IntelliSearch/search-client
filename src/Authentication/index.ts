export * from "./Authentication";
export * from "./AuthenticationSettings";
export * from "./AuthenticationTriggers";
export * from "./AuthToken";
