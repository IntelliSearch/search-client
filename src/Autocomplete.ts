import { AutocompleteSettings } from './AutocompleteSettings';

/**
 * Defines the parameters for the autocomplete() call.
 */
export class Autocomplete extends AutocompleteSettings {

    /**
     * The constructor can either take the listed params as function arguments, or you can send in a object with each of the listed params as keys (JSON notation).
     * 
     * @param queryText - The text to find autocomplete suggestions for. @default "".
     * @param maxSuggestions - The number of suggestions to return. @default 10.
     * @param minQueryLength - The minimum length before starting to create suggestions. @default 2.
     */
    constructor(queryText?: string | AutocompleteSettings, maxSuggestions?: number, minQueryLength?: number) {
        super();
        let o: AutocompleteSettings = typeof queryText === "object" ? queryText : {} as AutocompleteSettings;
        this.queryText = o.queryText || queryText as string || '';
        this.maxSuggestions = o.maxSuggestions || maxSuggestions || 10;
        this.minQueryLength = o.minQueryLength || minQueryLength || 2;
    }

    /**
     * Returns the specific rest-path segment for the autocomplete url.
     */
    public toUrlParam(): string {
        let params: string[] = [];

        params.push(`q=${encodeURIComponent(this.queryText)}`);
        params.push(`s=${encodeURIComponent(this.maxSuggestions.toString())}`);
        params.push(`l=${encodeURIComponent(this.minQueryLength.toString())}`);

        return `?${params.join('&')}`;
    }
}
