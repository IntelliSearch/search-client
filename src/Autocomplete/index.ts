export * from './AutocompleteTrigger';
export * from './AutocompleteSettings';
export * from './Autocomplete';
