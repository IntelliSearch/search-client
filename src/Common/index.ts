export * from './BaseCall';
export * from './BaseSettings';
export * from './BaseTriggers';
export * from './BaseQueryConverter';
export * from './CategorizationType';
export * from './Filter';
export * from './OrderBy';
export * from './Query';
export * from './SearchType';
