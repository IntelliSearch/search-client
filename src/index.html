<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <title>IntelliSearch Testbench</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style>
      .code {
        white-space: pre;
        background-color: darkslategray;
        color: white;
        margin: 0.5em 1em;
        padding: 0 0.5em;
        display: block;
      }
      .code::before{
        content: "> ";
      }
      a {
        font-weight: bold;
      }
      li {
        margin-bottom: 1em;
      }
      ol li{
        font-weight: bold;
      }
      </style>
  </head>

  <body>
    <h1>IntelliSearch SearchClient Testbench</h1>
    <p>
      In order to have the test-results live updated and nicely presented you need to:
      <ol>
        <li>Install the 'TAP Browser Chrome extension' (see below).</li>
        <li>Expose the test-data by running 'json-server' (see below).</li>
        <li>Run 'webpack-dev-server' (see below).</li>
        <li>Open the browser on http://localhost:8080</li>
        <li>Open the development console</li>
        <li>If on Chrome: Open the TAP tab (requires the Chrome extension mentioned above)</li>
        <li>For other browsers: Watch the console for test-results in TAP format (plain).
      </ol>
    </p>
    <h2>Tech notes:</h2>
    <ul>
      <li><b>Dev mode webpack:</b><br/>
        In order to execute tests for the search-client you will have to run:
        <span class="code">npm run dev</span>
        The files are then loaded through webpack and the results are served via this html-file. 
        Note that the tests are written as a part of the webpack entry point index.ts file. 
      </li>
      <li><b>Production mode webpack:</b><br/>
        To build the production search-client, run:
        <span class="code">npm run build</span>
        This will use the same entry point for webpack, but excludes the 'tape' test-references and
        ensures that the tests are not included in the production build.
      </li>
      <li><b>TAP Browser Chrome extension:</b><br/>
        The TAP results are plain and "boring" in the console. If you install thisfree extension 
        you will get a new TAP tab in the browser development-console which will display test-results 
        in a much nicer way. Install from the Chrome store here: <a href="https://chrome.google.com/webstore/detail/browser-tap/ncfblaiipckncgeipgmpdioedcdmofei">https://chrome.google.com/webstore/detail/browser-tap/ncfblaiipckncgeipgmpdioedcdmofei</a>
      </li>
      <li><b>Test-data.json as REST mockup data via "json-server":</b><br/>
        Some tests depend on test-data to be available to be successful. The test-data is 
        stored in the local test-date.json file. These json objects are RESTfully delivered 
        by <a href="https://github.com/typicode/json-server">json-server</a>.<br/>
        Start the server by running (from the root of the project): <span class="code">json-server --watch test-data.json --routes test-data-routes.json</span><br/>
        You will need to have json-server globally installed first: <span class="code">npm --global install json-server</span>
      </li>
      <li><b>About TAP tests:</b><br/>
        The search-client is setup to run it's test-suite and report results according to 
        the <a href="https://testanything.org/">TAP protocol</a>. The tests 
        are written using the TAP compliant <a href="https://github.com/substack/tape">tape</a> 
        npm module.
      </li>
    </ul>
    <div id="app"></div>
    <script src="/assets/search-client.js" type="text/javascript"></script>
  </body>
</html>
