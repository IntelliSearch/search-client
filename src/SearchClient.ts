import * as equal from 'deep-equal';
export * from './Common';
export * from './Data';
export * from './Authentication';
export * from './AllCategories';
export * from './Autocomplete';
export * from './BestBets';
export * from './Categorize';
export * from './Find';

import { OrderBy } from './Common/OrderBy';
import { SearchType } from './Common/SearchType';
import { Settings } from './Common/Settings';
import { DateSpecification, Query } from './Common/Query';

import { AuthToken } from './Authentication/AuthToken';
import { Authentication } from './Authentication/Authentication';

import { AllCategories } from './AllCategories/AllCategories';
import { Autocomplete } from './Autocomplete/Autocomplete';
import { AutocompleteSettings } from './Autocomplete/AutocompleteSettings';
import { AutocompleteTrigger } from './Autocomplete/AutocompleteTrigger';
import { BestBets } from './BestBets/BestBets';
import { Categorize } from './Categorize/Categorize';
import { Find } from './Find/Find';

export class SearchClient implements AuthToken {
    public authenticationToken: string;

    public allCategories: AllCategories;

    public authentication: Authentication;

    public autocomplete: Autocomplete;

    public bestBets: BestBets;

    public categorize: Categorize;
    
    public find: Find;

    private query: Query;

    private settings: Settings;
    
    constructor(baseUrl: string, settings?: Settings) {
        this.settings = new Settings(settings);

        if (this.settings.allCategories.enabled) {
            this.allCategories = new AllCategories(baseUrl, this.settings.allCategories, this);
        }

        if (this.settings.authentication.enabled) {
            this.authentication = new Authentication(baseUrl, this.settings.authentication, this);
        }

        if (this.settings.autocomplete.enabled) {
            this.autocomplete = new Autocomplete(baseUrl, this.settings.autocomplete, this);
        }

        if (this.settings.bestBets.enabled) {
            this.bestBets = new BestBets(baseUrl, this.settings.bestBets, this);
        }

        if (this.settings.categorize.enabled) {
            this.categorize = new Categorize(baseUrl, this.settings.categorize, this);
        }

        if (this.settings.find.enabled) {
            this.find = new Find(baseUrl, this.settings.find, this);
        }

        this.query = this.settings.query;
    }

    // public monitor(queryElm: Element) {
    //     throw Error("Not implemented yet!");
    // }

    get clientId(): string {
        return this.query.clientId;
    }

    set clientId(clientId: string) {
        if (clientId !== this.query.clientId) {
            let oldValue = this.query.clientId; 
            this.query.clientId = clientId;

            this.autocomplete.clientIdChanged(oldValue, this.query);
            this.categorize.clientIdChanged(oldValue, this.query);
            this.find.clientIdChanged(oldValue, this.query);
        }
    }

    get dateFrom(): DateSpecification {
        return this.query.dateFrom;
    }

    set dateFrom(dateFrom: DateSpecification) {
        if (!equal(dateFrom, this.query.dateFrom)) {
            let oldValue = Object.assign({}, this.query.dateFrom); // clone
            this.query.dateFrom = dateFrom;

            this.autocomplete.dateFromChanged(oldValue, this.query);
            this.categorize.dateFromChanged(oldValue, this.query);
            this.find.dateFromChanged(oldValue, this.query);
        }
    }

    get dateTo(): DateSpecification {
        return this.query.dateTo;
    }

    set dateTo(dateTo: DateSpecification) {
        if (!equal(dateTo, this.query.dateTo)) {
            let oldValue = Object.assign({}, this.query.dateTo); // clone
            this.query.dateTo = dateTo;

            this.autocomplete.dateToChanged(oldValue, this.query);
            this.categorize.dateToChanged(oldValue, this.query);
            this.find.dateToChanged(oldValue, this.query);
        }
    }

    get filters(): string[] {
        return this.query.filters;
    }

    set filters(filters: string[]) {
        filters = filters || [];
        let sortedFilters = filters.sort();
        if (sortedFilters.join('') !== this.query.filters.join('')) {
            let oldValue = this.query.filters.slice(0); // clone
            this.query.filters = sortedFilters;

            this.autocomplete.filtersChanged(oldValue, this.query);
            this.categorize.filtersChanged(oldValue, this.query);
            this.find.filtersChanged(oldValue, this.query);
        }
    }

    public filterAdd(filter: string): boolean {
        if (this.query.filters.indexOf(filter) === -1) {
            let oldValue = this.query.filters.slice(0);
            this.query.filters.push(filter);
            this.query.filters.sort();

            this.autocomplete.filtersChanged(oldValue, this.query);
            this.categorize.filtersChanged(oldValue, this.query);
            this.find.filtersChanged(oldValue, this.query);
            
            return true;
        } 
        // Filter already set
        return false;
    }

    public filterRemove(filter: string): boolean {
        const pos = this.query.filters.indexOf(filter);
        if (pos > -1) {
            let oldValue = this.query.filters.slice(0);
            this.query.filters.splice(pos, 1); 
            // Note: No need to sort the filter-list afterwards, as removing an item cannot change the order anyway.

            this.autocomplete.filtersChanged(oldValue, this.query);
            this.categorize.filtersChanged(oldValue, this.query);
            this.find.filtersChanged(oldValue, this.query);

            return true;
        } 
        // Filter already not set
        return false;
    }

    get matchGrouping(): boolean {
        return this.query.matchGrouping;
    }

    set matchGrouping(useGrouping: boolean) {
        if (useGrouping !== this.query.matchGrouping) {
            let oldValue = this.query.matchGrouping;
            this.query.matchGrouping = useGrouping;

            this.autocomplete.matchGroupingChanged(oldValue, this.query);
            this.categorize.matchGroupingChanged(oldValue, this.query);
            this.find.matchGroupingChanged(oldValue, this.query);
        }
    }

    get matchPage(): number {
        return this.query.matchPage;
    }

    set matchPage(page: number) {
        if (page < 0) {
            page = 0;
        }
        if (page !== this.query.matchPage) {
            let oldValue = this.query.matchPage;
            this.query.matchPage = page;

            this.autocomplete.matchPageChanged(oldValue, this.query);
            this.categorize.matchPageChanged(oldValue, this.query);
            this.find.matchPageChanged(oldValue, this.query);
        }
    }

    /**
     * This method is typically called when the user clicks the search-button in the UI.
     * For query-fields that accepts enter the default queryChangeInstantRegex catches enter (for find and categorize).
     * When they don't take enter you will have to set up something that either catches the default enter or a user clicks
     * on a "Search"-button or similar. The query is already updated (by other methods), so there should be no need to 
     * pass the query along, it should already be stored.
     * 
     * When called it will unconditionally call the fetch() method of both Categorize and Find.
     * 
     * Note: The Autocomplete fetch() method is not called, as it is deemed very unexpected to awnt to list autocomplete 
     * suggestions when the Search-button is clicked.
     */
    public go() {
        this.categorize.fetch(this.query);
        this.find.fetch(this.query);
    }

    public matchPagePrev(): boolean {
        if (this.query.matchPage > 0) {
            let oldValue = this.query.matchPage;
            this.query.matchPage--;

            this.autocomplete.matchPageChanged(oldValue, this.query);
            this.categorize.matchPageChanged(oldValue, this.query);
            this.find.matchPageChanged(oldValue, this.query);

            return true;
        }
        // Cannot fetch page less than 0
        return false;
    }

    public matchPageNext(): boolean {
        let oldValue = this.query.matchPage;
        this.query.matchPage++;

        this.autocomplete.matchPageChanged(oldValue, this.query);
        this.categorize.matchPageChanged(oldValue, this.query);
        this.find.matchPageChanged(oldValue, this.query);

        return true;
    }

    get matchPageSize(): number {
        return this.query.matchPageSize;
    }

    set matchPageSize(pageSize: number) {
        if (pageSize < 1) {
            pageSize = 1;
        }
        if (pageSize !== this.query.matchPageSize) {
            let oldValue = this.query.matchPageSize;
            this.query.matchPageSize = pageSize;

            this.autocomplete.matchPageSizeChanged(oldValue, this.query);
            this.categorize.matchPageSizeChanged(oldValue, this.query);
            this.find.matchPageSizeChanged(oldValue, this.query);
        }
    }

    get matchOrderBy(): OrderBy {
        return this.query.matchOrderBy;
    }

    set matchOrderBy(orderBy: OrderBy) {
        if (orderBy !== this.query.matchOrderBy) {
            let oldValue = this.query.matchOrderBy;
            this.query.matchOrderBy = orderBy;

            this.autocomplete.matchOrderByChanged(oldValue, this.query);
            this.categorize.matchOrderByChanged(oldValue, this.query);
            this.find.matchOrderByChanged(oldValue, this.query);
        }
    }

    get maxSuggestions(): number {
        return this.query.maxSuggestions;
    }

    set maxSuggestions(maxSuggestions: number) {
        if (maxSuggestions < 0) {
            maxSuggestions = 0;
        }
        if (maxSuggestions !== this.query.maxSuggestions) {
            let oldValue = this.query.maxSuggestions;
            this.query.maxSuggestions = maxSuggestions;

            this.autocomplete.maxSuggestionsChanged(oldValue, this.query);
            this.categorize.maxSuggestionsChanged(oldValue, this.query);
            this.find.maxSuggestionsChanged(oldValue, this.query);
        }
    }

    get queryText(): string {
        return this.query.queryText;
    }

    set queryText(queryText: string) {
        if (queryText !== this.query.queryText) {
            let oldValue = this.query.queryText;
            this.query.queryText = queryText;

            this.autocomplete.queryTextChanged(oldValue, this.query);
            this.categorize.queryTextChanged(oldValue, this.query);
            this.find.queryTextChanged(oldValue, this.query);
        }
    }

    get searchType(): SearchType {
        return this.query.searchType;
    }

    set searchType(searchType: SearchType) {
        if (searchType !== this.query.searchType) {
            let oldValue = this.query.searchType;
            this.query.searchType = searchType;

            this.autocomplete.searchTypeChanged(oldValue, this.query);
            this.categorize.searchTypeChanged(oldValue, this.query);
            this.find.searchTypeChanged(oldValue, this.query);
        }
    }
}
