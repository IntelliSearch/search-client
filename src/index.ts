const test = require('tape').default;

if(process.env.NODE_ENV === 'ci') {
    test.onFinish(function(){
        window.close();
    });
}

import SearchClient, { Query } from './search-client';

test('Test initialization:', function (assert:any) {

    assert.throws(function(){
        let searchClient = new SearchClient("");
    }, 'Does not allow empty baseServiceUrl.');

    assert.throws(function(){
        let searchClient = new SearchClient("http://myaccount.intellisearch:com");
    }, 'Should not allow for invalid url constructs (http://myaccount.intellisearch:com).');

    assert.doesNotThrow(function(){
        let searchClient = new SearchClient("http://myaccount.intellisearch.com");
        assert.equal(searchClient.allCategoriesUrl.toString(), "http://myaccount.intellisearch.com/search/allcategories", 'The allCategoriesUrl is as expected.')
        assert.equal(searchClient.autocompleteUrl.toString(), "http://myaccount.intellisearch.com/autocomplete", 'The autocompleteUrl is as expected.')
        assert.equal(searchClient.bestBetsUrl.toString(), "http://myaccount.intellisearch.com/manage/bestbets", 'The bestBetsUrl is as expected.')
        assert.equal(searchClient.categorizeUrl.toString(), "http://myaccount.intellisearch.com/search/categorize", 'The categorizeUrl is as expected.')
        assert.equal(searchClient.findUrl.toString(), "http://myaccount.intellisearch.com/search/find", 'The findUrl is as expected.')
    }, 'Should allow with just the baseServiceUrl.');
    
    assert.end();
});

test('Test find(), with callback from settings', function (assert:any) {
    assert.doesNotThrow(function(){
        let searchClient = new SearchClient("http://localhost:3000", {
            callback: {
                find: (matches) => {
                    assert.equal(matches.estimatedMatchCount, 101, "Did receive 101 matches in the settings find callback-handler.");
                }
            }
        });
        searchClient.find(new Query())
        .then((matches)=>{
            assert.equal(matches.estimatedMatchCount, 101, "Did receive 101 matches as a resolved promise.");
            assert.end();
            return Promise.resolve();
        })
        .catch((reason)=>{
            assert.fail(`${reason}`);
            assert.end();
            return Promise.resolve();
        });
    }, "Doesn't throw an exception.");
});

test('Test categorize(), with callback from settings', function (assert:any) {
    assert.doesNotThrow(function(){
        let searchClient = new SearchClient("http://localhost:3000", {
            callback: {
                categorize: (matches) => {
                    assert.equal(matches.matchCount, 101, "Did receive 101 matches in the settings categorize callback-handler.");
                }
            }
        });
        searchClient.categorize(new Query)
        .then((categories)=>{
            assert.equal(categories.matchCount, 101, "Did receive 101 matches as a resolved promise.");
            assert.end();
            return Promise.resolve();
        })
        .catch((reason)=>{
            assert.fail(`${reason}`);
            assert.end();
            return Promise.resolve();
        });
    }, "Doesn't throw an exception.");
});

test('Test find(), with override callback', function (assert:any) {
    assert.doesNotThrow(function(){
        let searchClient = new SearchClient("http://localhost:3000", {
            callback: {
                find: (matches) => {
                    assert.fail("Should not call the settings find callback-handler.");
                }
            }
        });
        searchClient.find(new Query(), (matches) => { 
            assert.equal(matches.estimatedMatchCount, 101, "Did receive 101 matches in the override find callback-handler.") 
        })
        .then((matches)=>{
            assert.equal(matches.estimatedMatchCount, 101, "Did receive 101 matches as a resolved promise.");
            assert.end();
            return Promise.resolve();
        })
        .catch((reason)=>{
            assert.fail(`${reason}`);
            assert.end();
            return Promise.resolve();
        });
    }, "Doesn't throw an exception.");
});

test('Test categorize(), with override callback', function (assert:any) {
    assert.doesNotThrow(function(){
        let searchClient = new SearchClient("http://localhost:3000", {
            callback: {
                categorize: (categories) => {
                    assert.fail("Should not call the settings categorize callback-handler.");
                }
            }
        });
        searchClient.categorize(new Query, (categories) => { 
            assert.equal(categories.matchCount, 101, "Did receive 101 matches in the override categorize callback-handler.") 
        })
        .then((categories)=>{
            assert.equal(categories.matchCount, 101, "Did receive 101 matches as a resolved promise.");
            assert.end();
            return Promise.resolve();
        }).catch((reason)=>{
            assert.fail(`${reason}`);
            assert.end();
            return Promise.resolve();
        });
    }, "Doesn't throw an exception.");
});

test('Test find(), with override callback only', function (assert:any) {
    assert.doesNotThrow(function(){
        let searchClient = new SearchClient("http://localhost:3000");
        searchClient.find(new Query(), (matches) => { 
            assert.equal(matches.estimatedMatchCount, 101, "Did receive 101 matches in the override find callback-handler.");
            assert.end();
        }).catch((reason)=>{
            assert.fail(`${reason}`);
            assert.end();
            return Promise.resolve();
        });
    }, "Doesn't throw an exception.");
});

test('Test categorize(), with override callback only', function (assert:any) {
    assert.doesNotThrow(function(){
        let searchClient = new SearchClient("http://localhost:3000");
        searchClient.categorize(new Query, (categories) => { 
            assert.equal(categories.matchCount, 101, "Did receive 101 matches in the override categorize callback-handler.");
            assert.end();
        })
        .catch((reason)=>{
            assert.fail(`${reason}`);
            assert.end();
            return Promise.resolve();
        });
    }, "Doesn't throw an exception.");
});

test('Test Query to*UrlParam() methods', function (assert:any) {
    assert.doesNotThrow(function(){
        let query = new Query();
        query.queryText = "test";
        query.searchType = 1;
        query.filters = ["System/Exchange/Mailbox", "System/File/Fileshare1"];
        query.dateFrom = new Date(2016, 1, 1);
        query.dateTo = new Date(2017, 1, 27);
        query.clientId = "mobile";
        query.pageSize = 5;
        query.page = 2;
        query.useGrouping = true;
        query.orderBy = 1;
        assert.equal(query.toFindUrlParam(), "?q=test&t=Relevance&f=System/Exchange/Mailbox;System/File/Fileshare1&df=2016-01-31T23:00:00.000Z&dt=2017-02-26T23:00:00.000Z&c=mobile&s=5&p=2&g=true&o=Relevance", "Query is serialized as url param for find() as expected.");
        assert.equal(query.toCategorizeUrlParam(), "?q=test&t=Relevance&f=System/Exchange/Mailbox;System/File/Fileshare1&df=2016-01-31T23:00:00.000Z&dt=2017-02-26T23:00:00.000Z&c=mobile", "Query is serialized as url param for categorize() as expected.");
        let searchClient = new SearchClient("http://localhost:3000");
        searchClient.find(query, (matches) => { 
            assert.equal(matches.estimatedMatchCount, 101, "Did receive 101 matches in the override find callback-handler.");
        })
        .then((matches)=>{
            assert.equal(matches.estimatedMatchCount, 101, "Did receive 101 matches as a resolved promise.");
            assert.end;
        })
        .catch((reason)=>{
            assert.fail(`${reason}`);
            assert.end();
            return Promise.resolve();
        });
    }, "Doesn't throw an exception.");
});
