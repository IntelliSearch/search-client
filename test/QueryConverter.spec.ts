import { OrderBy, SearchType } from '../src/Common';
// tslint:disable-next-line:no-var-requires
require("babel-core/register");
require("babel-polyfill");

import { Query, QueryConverter, QueryCategorizeConverterV2, QueryCategorizeConverterV3, QueryFindConverterV2, QueryFindConverterV3 } from '../src/SearchClient';

const query = new Query({
    clientId: "mobile", 
    dateFrom: "2017-03-13 09Z", 
    dateTo: "2017-03-13 09Z", 
    filters: ["Authors|Bob", "FileTypes|docx"], 
    matchGrouping: true, 
    matchOrderBy: OrderBy.Date, 
    matchPage: 1, 
    matchPageSize: 20, 
    maxSuggestions: 20, 
    queryText: "test", 
    searchType: SearchType.Keywords,
});

describe("QueryConverters", () => {

    it("Should have QueryConverter interface", () => {
        let cc2 = (<any> new QueryCategorizeConverterV2());
        let cc3 = (<any> new QueryCategorizeConverterV3());
        let cf2 = (<any> new QueryFindConverterV2());
        let cf3 = (<any> new QueryFindConverterV3());

        expect((cc2 as QueryConverter).getUrl).toBeDefined();
    });

    it("Should match expectations for REST V2", () => {
        let cc2 = (<any> new QueryCategorizeConverterV2());
        let cf2 = (<any> new QueryFindConverterV2());
        const expectedCatUrl = "http://localhost:9950/RestService/v2/search/categorize/test?c=mobile&f=Authors%7CBob%3BFileTypes%7Cdocx&t=Keywords";
        expect(cc2.getUrlParams(query)).toHaveLength(3);
        expect(cc2.getUrl("http://localhost:9950/RestService/v2", "search/categorize", query)).toEqual(expectedCatUrl);
        expect(cc2.getUrl("http://localhost:9950/RestService/v2/", "search/categorize", query)).toEqual(expectedCatUrl);
        expect(cc2.getUrl("http://localhost:9950/RestService/v2", "search/categorize/", query)).toEqual(expectedCatUrl);
        expect(cc2.getUrl("http://localhost:9950/RestService/v2/", "search/categorize/", query)).toEqual(expectedCatUrl);
        expect(cc2.getUrl("http://localhost:9950/RestService/v2", "/search/categorize/", query)).toEqual(expectedCatUrl);
        expect(cc2.getUrl("http://localhost:9950/RestService/v2/", "/search/categorize/", query)).toEqual(expectedCatUrl);

        const expectedFindUrl = "http://localhost:9950/RestService/v2/search/find/test?c=mobile&f=Authors%7CBob%3BFileTypes%7Cdocx&g=true&o=Date&p=1&s=20&t=Keywords";
        expect(cf2.getUrlParams(query)).toHaveLength(7);
        expect(cf2.getUrl("http://localhost:9950/RestService/v2", "search/find", query)).toEqual(expectedFindUrl);
        expect(cf2.getUrl("http://localhost:9950/RestService/v2/", "search/find", query)).toEqual(expectedFindUrl);
        expect(cf2.getUrl("http://localhost:9950/RestService/v2", "search/find/", query)).toEqual(expectedFindUrl);
        expect(cf2.getUrl("http://localhost:9950/RestService/v2/", "search/find/", query)).toEqual(expectedFindUrl);
        expect(cf2.getUrl("http://localhost:9950/RestService/v2", "/search/find/", query)).toEqual(expectedFindUrl);
        expect(cf2.getUrl("http://localhost:9950/RestService/v2/", "/search/find/", query)).toEqual(expectedFindUrl);
    });

    it("Should match expectations for REST V3", () => {
        let cc3 = (<any> new QueryCategorizeConverterV3());
        let cf3 = (<any> new QueryFindConverterV3());

        const expectedCatUrl = "http://localhost:9950/RestService/v3/search/categorize?c=mobile&df=2017-03-13T09%3A00%3A00.000Z&dt=2017-03-13T09%3A00%3A00.000Z&f=Authors%7CBob%3BFileTypes%7Cdocx&q=test&t=Keywords";
        expect(cc3.getUrlParams(query)).toHaveLength(6);
        expect(cc3.getUrl("http://localhost:9950/RestService/v3", "search/categorize", query)).toEqual(expectedCatUrl);
        expect(cc3.getUrl("http://localhost:9950/RestService/v3/", "search/categorize", query)).toEqual(expectedCatUrl);
        expect(cc3.getUrl("http://localhost:9950/RestService/v3", "search/categorize/", query)).toEqual(expectedCatUrl);
        expect(cc3.getUrl("http://localhost:9950/RestService/v3/", "search/categorize/", query)).toEqual(expectedCatUrl);
        expect(cc3.getUrl("http://localhost:9950/RestService/v3", "/search/categorize/", query)).toEqual(expectedCatUrl);
        expect(cc3.getUrl("http://localhost:9950/RestService/v3/", "/search/categorize/", query)).toEqual(expectedCatUrl);

        const expectedFindUrl = "http://localhost:9950/RestService/v3/search/find?c=mobile&df=2017-03-13T09%3A00%3A00.000Z&dt=2017-03-13T09%3A00%3A00.000Z&f=Authors%7CBob%3BFileTypes%7Cdocx&g=true&o=Date&p=1&q=test&s=20&t=Keywords";
        expect(cf3.getUrlParams(query)).toHaveLength(10);
        expect(cf3.getUrl("http://localhost:9950/RestService/v3", "search/find", query)).toEqual(expectedFindUrl);
        expect(cf3.getUrl("http://localhost:9950/RestService/v3/", "search/find", query)).toEqual(expectedFindUrl);
        expect(cf3.getUrl("http://localhost:9950/RestService/v3", "search/find/", query)).toEqual(expectedFindUrl);
        expect(cf3.getUrl("http://localhost:9950/RestService/v3/", "search/find/", query)).toEqual(expectedFindUrl);
        expect(cf3.getUrl("http://localhost:9950/RestService/v3", "/search/find/", query)).toEqual(expectedFindUrl);
        expect(cf3.getUrl("http://localhost:9950/RestService/v3/", "/search/find/", query)).toEqual(expectedFindUrl);
    });

});
