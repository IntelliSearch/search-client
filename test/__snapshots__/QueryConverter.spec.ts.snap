// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryConverters Should match expectations for REST V2 1`] = `"baseUrl/test?c=mobile&f=Authors%7CBob%3BFileTypes%7Cdocx&t=Keywords"`;

exports[`QueryConverters Should match expectations for REST V2 2`] = `"baseUrl/test?c=mobile&f=Authors%7CBob%3BFileTypes%7Cdocx&t=Keywords"`;

exports[`QueryConverters Should match expectations for REST V2 3`] = `"baseUrl/test?c=mobile&f=Authors%7CBob%3BFileTypes%7Cdocx&g=true&o=Date&p=1&s=20&t=Keywords"`;

exports[`QueryConverters Should match expectations for REST V2 4`] = `"baseUrl/test?c=mobile&f=Authors%7CBob%3BFileTypes%7Cdocx&g=true&o=Date&p=1&s=20&t=Keywords"`;

exports[`QueryConverters Should match expectations for REST V3 1`] = `"baseUrl?c=mobile&df=2017-03-13T09%3A00%3A00.000Z&dt=2017-03-13T09%3A00%3A00.000Z&f=Authors%7CBob%3BFileTypes%7Cdocx&q=test&t=Keywords"`;

exports[`QueryConverters Should match expectations for REST V3 2`] = `"baseUrl?c=mobile&df=2017-03-13T09%3A00%3A00.000Z&dt=2017-03-13T09%3A00%3A00.000Z&f=Authors%7CBob%3BFileTypes%7Cdocx&q=test&t=Keywords"`;

exports[`QueryConverters Should match expectations for REST V3 3`] = `"baseUrl?c=mobile&df=2017-03-13T09%3A00%3A00.000Z&dt=2017-03-13T09%3A00%3A00.000Z&f=Authors%7CBob%3BFileTypes%7Cdocx&g=true&o=Date&p=1&q=test&s=20&t=Keywords"`;

exports[`QueryConverters Should match expectations for REST V3 4`] = `"baseUrl?c=mobile&df=2017-03-13T09%3A00%3A00.000Z&dt=2017-03-13T09%3A00%3A00.000Z&f=Authors%7CBob%3BFileTypes%7Cdocx&g=true&o=Date&p=1&q=test&s=20&t=Keywords"`;
